### PR TITLE
Improve diagnostics to ease debugging

### PR DESCRIPTION
**Pull request by krafczyk**
The option -s now causes file and line number information to be printed
along with any invocation of msg, info, etc...

This will greatly ease debugging.

### CODE REVIEW

**Comment by krafczyk** [comment]
Example:

```
spack -s -d -v checksum py-matplotlib
lib/spack/spack/util/web.py:146 ==> HTTP Error 404: Not Found
lib/spack/spack/cmd/checksum.py:92 ==> Error: Could not fetch any versions for py-matplotlib
```

There is now no question where those print statements are coming from for new developers.

File: lib/spack/llnl/util/tty/__init__.py [hunk 1]
```diff
@@ -53,10 +59,29 @@ def set_debug(flag):
...
+
+
+def process_stacktrace(countback):
+    """Gives file and line frame 'countback' frames from the bottom"""
+    st = traceback.extract_stack()
+    # First entry should be bin/spack. Use this to get spack 'root'.
+    # bin/spack is 9 characters, the length of the 'root' is then len-9.
+    root_len = len(st[0][0]) - 9
+    st_idx = len(st) - countback - 1
+    st_text = "%s:%i " % (st[st_idx][0][root_len:], st[st_idx][1])
```
Comment (adamjstewart) [inline lib/spack/llnl/util/tty/__init__.py]: This seems kind of hacky. If you're trying to get the root directory of Spack, why not use `os.path.basename` a couple times instead?

Comment (krafczyk) [inline lib/spack/llnl/util/tty/__init__.py]: I've updated and and rebased the branch, and added a better method for getting the spack root directory.

All files in the stacktrace will be spack files and hence we can just use os.path.commonprefix to fetch the commonprefix.

Comment (adamjstewart) [inline lib/spack/llnl/util/tty/__init__.py]: Never used `os.path.commonprefix` before. Very interesting!

**Comment by krafczyk** [comment]
I don't know why the test failed.. Maybe this is an instance of #2353?

**Comment by adamjstewart** [comment]
It is indeed. I added it to the list. If you close and reopen this PR it will re-run Travis.

**Comment by krafczyk** [comment]
@tgamblin, is it possible to get this merged?

**Pull request merged**